zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -132,14 +132,34 @@ pub(crate) fn get_spendable_sapling_notes(
     anchor_height: BlockHeight,
     exclude: &[ReceivedNoteId],
 ) -> Result<Vec<ReceivedSaplingNote<ReceivedNoteId>>, SqliteClientError> {
+    let mut stmt_unscanned_tip = conn.prepare_cached(
+        "SELECT 1 FROM v_sapling_shard_unscanned_ranges unscanned
+         WHERE unscanned.shard_index IN (SELECT MAX(shard_index) FROM sapling_tree_shards)
+         AND unscanned.subtree_start_height <= :anchor_height",
+    )?;
+    let mut unscanned =
+        stmt_unscanned_tip.query(named_params![":anchor_height": &u32::from(anchor_height),])?;
+    if unscanned.next()?.is_some() {
+        // if the tip shard has unscanned ranges below the anchor height, none of our notes can be
+        // spent
+        return Ok(vec![]);
+    }
+
     let mut stmt_select_notes = conn.prepare_cached(
         "SELECT id_note, diversifier, value, rcm, commitment_tree_position
          FROM sapling_received_notes
          INNER JOIN transactions ON transactions.id_tx = sapling_received_notes.tx
          WHERE account = :account
+         AND commitment_tree_position IS NOT NULL
          AND spent IS NULL
          AND transactions.block <= :anchor_height
-         AND id_note NOT IN rarray(:exclude)",
+         AND id_note NOT IN rarray(:exclude)
+         AND NOT EXISTS (
+            SELECT 1 FROM v_sapling_shard_unscanned_ranges unscanned
+            WHERE sapling_received_notes.commitment_tree_position >= unscanned.start_position
+            AND sapling_received_notes.commitment_tree_position < unscanned.end_position_exclusive
+            AND unscanned.block_range_start <= :anchor_height
```

Comment:
I think the correct condition here is:
```suggestion
            WHERE (
                sapling_received_notes.commitment_tree_position >= unscanned.start_position
                AND sapling_received_notes.commitment_tree_position < unscanned.end_position_exclusive
            ) OR (
                sapling_received_notes.commitment_tree_position < unscanned.start_position
                AND unscanned.block_range_start <= :anchor_height
            )
```

Reply:
```
sapling_received_notes.commitment_tree_position < unscanned.start_position
AND unscanned.block_range_start <= :anchor_height
```
This seems to select too many ranges; it can include unscanned subtrees for which we have the root, and therefore we don't need the range to be scanned.

Reply:
Ah, I was misreading the new view in that case. That means AFAICT we cannot filter out the second conditional at all, and instead have to manually check whether we can obtain witnesses after the notes have been selected.

It remains the case that the first conditional above is valid, i.e. `AND transactions.block <= :anchor_height` at the note level implies that `AND unscanned.block_range_start <= :anchor_height` is redundant at the `NOT EXISTS` conditional.

Reply:
In order for a note to be spendable, two subtrees have to be fully scanned: the subtree containing the note, and the portion of the subtree at the chain tip up to and including the anchor height. The latter is the same for all notes, so we ensure that it has been scanned using the query on lines 136..139; if that range hasn't been scanned, then no notes are spendable and we can exit early.

The new view has one entry per scan range (not one entry per shard) and so we want the `EXISTS` subquery to find any unscanned range that overlaps with the range of blocks associated with the subtree containing the note being selected. The existing implementation, combined with the check related to 136..139 gives us exactly the notes where we have enough information to spend them.

Given this, I don't see why you say we would need to manually check whether we can construct the witnesses after retrieval?

Reply:
Comments have been added to the query to more clearly describe what's going on here.